### ISSUE TITLE
Generating MLIR code with dynamic dimensions via Reactant?

### ISSUE THREAD

**Issue by oschulz** [EnzymeAD/Reactant.jl#2809]
Reactant may be an awesome way to export, share and preserve scientific models (not just ML, also simulation code, etc.) written in Julia, so they can be used beyond Julia.

One problem I've run into while exploring this is how to generate StableHLO code that doesn't hardcode batch size. For example, when doing a multi-particle simulation, the exported StableHLO shouldn't hardcode the number of particles - batch size should still be flexible depending on user need and target hardware.

StableHLO itself support things like `tensor<?x128xf32>` right? And maybe the `?` could then even be rewritten to a fixed size at the point/time of use? I've read that JAX `export` supports emitting StableHLO with `?` dimensions via `polymorphic_shapes`, but I'm not an expert. Can/could Reactant support this as well?

This could be very relevant for a new project that I'm part of.

**Comment by mofeing** [comment]
Reactant arrays do not yet support dynamic dimensions, but we are discussing how they could get supported. Keep in mind that currently many optimizations require static shapes, so arrays with dynamic dimensions won't be as optimized as the ones with static shapes.

**Comment by oschulz** [comment]
>  but we are discussing how they could get supported.

Thanks, that would be great!

> Keep in mind that currently many optimizations require static shapes

Reactant or XLA optimizations? For the latter, it could be feasible to replace `?` by a batch size before use, at least in theory, right? Not by Reactant, but by whatever custom tooling that will read the stored StableHLO, before handing it to XLA?

**Comment by mofeing** [comment]
> Reactant or XLA optimizations?

Enzyme-JAX; i.e. the optimization engine under Reactant. I'm unaware of the status of XLA optimizations for dynamic dims.

> For the latter, it could be feasible to replace `?` by a batch size before use, at least in theory, right? Not by Reactant, but by whatever custom tooling that will read the stored StableHLO, before handing it to XLA?

Do you mean to compile it first with dyn dim and later transform the code with a given dim value? You would need to check if the value is propagated everywhere.

What are you trying to do?

**Comment by oschulz** [comment]
I'm trying to - well, more like exploring the feasibility of - preserving and exchanging physics model code in a host-language independent form. But with a batch size that is not fixed, but can be set at the "receiving end", so it speak. So basically I'm looking for a way to have an "insert batch size here" placeholder in MLIR/StableHLO code. I think typically there will be a fixed batch size at the end, at the point of use, and then fancy optimizations can be run by whatever framework is used there. Since all other dimension sizes will be fixed, I thought the dynamic dimension size of StableHLO could then, by convention, be used as the batch size - as there is a legal way to represent that in StableHLO code via `?` (from what I understand).

**Comment by avik-pal** [comment]
There are shape refinement passes implemented upstream in stablehlo https://openxla.org/stablehlo/dynamism#compiler_passes_for_refining_dynamic_programs

**Comment by yeruouo** [comment]
> There are shape refinement passes implemented upstream in stablehlo https://openxla.org/stablehlo/dynamism#compiler_passes_for_refining_dynamic_programs

Would it be beneficial to expose stablehlo-opt as a JLL package for the Julia ecosystem? From my understanding, it operates on MLIR input/output as strings, which might make it a good candidate for wrapping.

**Comment by wsmoses** [comment]
all of the stablehlo passes are available within the current reactant_jll [though as an api rather than standalone binary]. given how fast things move in mlir land it may not be as helpful to have both, in case one gets out of date from the other.

That said, @yeruoforever did you have a particular reason for a standalone executable rather than library as a jll? (also fun fact stablehlo-opt as well as many more opts are availalable within the mlir pipeline of enzyme.mit.edu/explorer)

**Comment by yeruouo** [comment]
I need to convert MLIR code with dynamic dimensions into MLIR code with static dimensions.

Thanks to the Reactant and Enzyme projects! Last weekend, I successfully exported a PyTorch model from Hugging Face via StableHLO and ran it on the Julia side. (To my excitement, for 100 batches of forward computation, the Julia side was 10 seconds faster than `torch.compile` — RTX 3090 GPU, sd-vae-ft.)

I successfully used Enzyme to compute gradients, which means I can declare the model architecture and forward computation in PyTorch, and then use Julia for training. This allows Reactant to serve as a bridge, connecting the Python and Julia deep learning ecosystems.

I'm polishing my exploratory scripts and found that when exporting the computation graph from PyTorch, dynamic dimensions are supported. However, in my attempts, I discovered that `Reactant.Ops.hlo_call` does not handle dynamic dimensions. I use a small batch size to quickly export the computation graph, and then use a larger batch size for training on the Julia side. Without dynamic shapes implemented, I have to go back to the Python side to re-adjust the batch size and re-export.

As discussed above, fixed dimensions are more favorable for optimization. In my scenario, Julia handles the computation, so the actual dimensions are determined at runtime. Naturally, I would like to perform the optimization on the Julia side, so I am looking for a tool or utility to do this.

Perhaps I am not familiar enough with Reactant (unable to implement this myself), which led me astray. I was previously hoping to use `stablehlo-opt` to convert MLIR code with dynamic dimensions into MLIR code with static dimensions.

Alternatively, are there any example codes based on `reactant_jll` for handling dynamic dimensions that I could refer to?
